File: apps/ios/ExpUI/Sources/EntityChip.swift
import ExpCore
import SwiftUI

// EXP-920 — THE entity chip, presentational half. An Exponential MCP tool's
// settled transcript row names what its answer touched (`preview.refs`,
// contract `entityRefKind`), and every client draws ONE chip per ref: a glyph
// and a short label in the very box the issue chip owns. `IssueChip` renders
// THROUGH `ChipBox` below with its three parts, so an issue chip and a board
// chip on the same row cannot disagree by a pixel.
//
// The glyph is handed IN: an issue passes its resolved status glyph, every
// other kind the concept `EntityPreview.refIcon` names (`EntityChipIcon`); a
// chip stays free of live queries either way. The label and its optional
// detail are the contract's (`EntityPreview.chipLabel` / `chipDetail`),
// fixture-locked ×4.

/// Concept → registry glyph for the entity kinds. `AppIcons` is generated as
/// one static member per concept, not a lookup table, so the entity kinds
/// name theirs here — `EntityChipTests` locks that every contract kind's
/// concept resolves to a shipped imageset.
public enum EntityChipIcon {
    public static let glyphs: [String: String] = [
        "ui-issue": AppIcons.uiIssue,
        "nav-boards": AppIcons.navBoards,
        "nav-actions": AppIcons.navActions,
        "nav-automations": AppIcons.navAutomations,
        "notification-issue-comment": AppIcons.notificationIssueComment,
        "coding-running": AppIcons.codingRunning,
        "settings-labels": AppIcons.settingsLabels,
        "settings-statuses": AppIcons.settingsStatuses,
        "nav-workflows": AppIcons.navWorkflows,
        "ui-device": AppIcons.uiDevice,
        "ui-avatar-placeholder": AppIcons.uiAvatarPlaceholder,
        "ui-repository": AppIcons.uiRepository,
        "ui-team": AppIcons.uiTeam,
        "ui-invite": AppIcons.uiInvite,
        "nav-notifications": AppIcons.navNotifications,
        "nav-support": AppIcons.navSupport,
        "ui-attach": AppIcons.uiAttach,
        "ui-checklist": AppIcons.uiChecklist,
    ]

    /// The glyph a concept draws; an unknown concept wears the list glyph,
    /// like an unknown kind does.
    public static func glyph(forConcept concept: String) -> String {
        glyphs[concept] ?? AppIcons.uiChecklist
    }

    /// The glyph a ref draws (`EntityPreview.refIcon`).
    public static func glyph(for ref: EntityRef) -> String {
        glyph(forConcept: EntityPreview.refIcon(ref))
    }
}

/// The shared chip box (EXP-885 paint, EXP-920 shape): a small rounded RECT
/// from the markdown chip's own tokens, a hairline border, and — left to
/// right — an optional leading glyph, an optional mono muted `identifier`, an
/// optional medium `title` and an optional trailing ✕ glyph. Every chip
/// outside the text painter is this box; the callers only decide the parts.
public struct ChipBox<Leading: View>: View {
    let identifier: String?
    let title: String?
    let titleColor: Color
    let bodySize: CGFloat?
    let showsRemoveGlyph: Bool
    @ViewBuilder let leading: () -> Leading

    public init(
        identifier: String?,
        title: String?,
        titleColor: Color = Color(MarkdownStyle.textColor),
        bodySize: CGFloat? = nil,
        showsRemoveGlyph: Bool = false,
        @ViewBuilder leading: @escaping () -> Leading
    ) {
        self.identifier = identifier
        self.title = title
        self.titleColor = titleColor
        self.bodySize = bodySize
        self.showsRemoveGlyph = showsRemoveGlyph
        self.leading = leading
    }

    private var pointSize: CGFloat { MarkdownStyle.resolvedBodyFont(bodySize).pointSize }

    public var body: some View {
        HStack(spacing: IssueChipTokens.spacing) {
            leading()
            if let identifier, !identifier.isEmpty {
                Text(identifier)
                    .font(.system(size: pointSize, design: .monospaced))
                    .foregroundStyle(Color(MarkdownStyle.chipTokenColor))
                    .lineLimit(1)
            }
            if let title, !title.isEmpty {
                Text(title)
                    .font(.system(size: pointSize, weight: .medium))
                    .foregroundStyle(titleColor)
                    .lineLimit(1)
                    .truncationMode(.tail)
                    .frame(maxWidth: IssueChipTokens.titleMaxWidth, alignment: .leading)
            }
            if showsRemoveGlyph {
                AppIcon(AppIcons.uiClose, size: IssueChipTokens.removeGlyphSize)
                    .foregroundStyle(Color(MarkdownStyle.chipTokenColor))
            }
        }
        .padding(.horizontal, IssueChipTokens.horizontalPadding)
        .padding(.vertical, IssueChipTokens.verticalPadding)
        .background(
            RoundedRectangle(cornerRadius: MarkdownStyle.chipCornerRadius, style: .continuous)
                .fill(Color(MarkdownStyle.chipBackground))
        )
        .overlay(
            RoundedRectangle(cornerRadius: MarkdownStyle.chipCornerRadius, style: .continuous)
                .strokeBorder(
                    Color(MarkdownStyle.chipBorder), lineWidth: IssueChipTokens.borderWidth)
        )
    }
}

/// One entity chip: a glyph, the contract's label and — for an issue that
/// carries both — the identifier as the mono lead-in with the title beside
/// it, exactly the `IssueChip` look. `muted` is the unsynced row: the same
/// box with its text in the token colour and, per the caller, no target.
public struct EntityChip: View {
    private let label: String
    private let detail: String?
    private let iconName: String?
    private let iconColor: Color?
    private let bodySize: CGFloat?
    private let muted: Bool
    private let onTap: (() -> Void)?

    /// - Parameters:
    ///   - label: the chip's text (`EntityPreview.chipLabel`).
    ///   - detail: the secondary text (`EntityPreview.chipDetail`); when it is
    ///     present the label reads as the mono identifier and the detail as
    ///     the title.
    ///   - iconName: the registry glyph (a resolved status glyph for a synced
    ///     issue, else `EntityChipIcon.glyph(for:)`); nil draws none.
    ///   - iconColor: the glyph's tint; nil = the muted token colour.
    ///   - bodySize: the prose measure the chip sits in (the transcript reads
    ///     at `DesignTokens.Transcript.bodySize`).
    public init(
        label: String,
        detail: String? = nil,
        iconName: String?,
        iconColor: Color? = nil,
        bodySize: CGFloat? = nil,
        muted: Bool = false,
        onTap: (() -> Void)? = nil
    ) {
        self.label = label
        self.detail = detail
        self.iconName = iconName
        self.iconColor = iconColor
        self.bodySize = bodySize
        self.muted = muted
        self.onTap = onTap
    }

    /// The contract convenience: label, detail and glyph straight off the ref.
    public init(
        ref: EntityRef,
        iconName: String? = nil,
        iconColor: Color? = nil,
        bodySize: CGFloat? = nil,
        muted: Bool = false,
        onTap: (() -> Void)? = nil
    ) {
        self.init(
            label: EntityPreview.chipLabel(ref),
            detail: EntityPreview.chipDetail(ref),
            iconName: iconName ?? EntityChipIcon.glyph(for: ref),
            iconColor: iconColor,
            bodySize: bodySize,
            muted: muted,
            onTap: onTap
        )
    }

    private var accessibilityText: String {
        [label, detail].compactMap { $0 }.joined(separator: " ")
    }

    public var body: some View {
        if let onTap {
            Button(action: onTap) { surface.contentShape(Rectangle()) }
                .buttonStyle(.plain)
                .accessibilityLabel(accessibilityText)
        } else {
            surface
                .accessibilityElement(children: .ignore)
                .accessibilityLabel(accessibilityText)
        }
    }

    private var surface: some View {
        ChipBox(
            identifier: detail == nil ? nil : label,
            title: detail ?? label,
            titleColor: muted ? Color(MarkdownStyle.chipTokenColor) : Color(MarkdownStyle.textColor),
            bodySize: bodySize
        ) {
            if let iconName {
                AppIcon(iconName, size: MarkdownStyle.chipStatusIconSize)
                    .foregroundStyle(iconColor ?? Color(MarkdownStyle.chipTokenColor))
            }
        }
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/components/EntityChip.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.markdown.MdStyle

/**
 * EXP-920: the entity badge — the SAME chip as [IssueChip] (one [ChipShell])
 * for every other kind an Exponential tool's answer names: a board, an
 * action, a run, a label… A glyph, then the label. An ISSUE ref with an
 * identifier reads exactly like [IssueChip]: the identifier in muted mono
 * ([monoLabel]) and the title as the medium [detail] beside it; anything else
 * is glyph + label in the body colour.
 *
 * [icon] is a slot, not a glyph: an issue paints its resolved status glyph
 * once its row is synced, a board its coloured board icon, a label its dot.
 */
@Composable
fun EntityChip(
    icon: @Composable () -> Unit,
    label: String,
    detail: String?,
    onClick: (() -> Unit)?,
    modifier: Modifier = Modifier,
    /** Draw [label] in the identifier style (mono, muted) rather than the
     *  body style — an issue named by its identifier. */
    monoLabel: Boolean = false,
) {
    ChipShell(modifier = modifier, onClick = onClick) {
        icon()
        Text(
            label,
            style = IssueChipDefaults.textStyle().let {
                if (monoLabel) it.copy(fontFamily = FontFamily.Monospace) else it
            },
            color = if (monoLabel) MdStyle.ChipToken else MdStyle.Text,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
        )
        if (detail != null) {
            Text(
                detail,
                style = IssueChipDefaults.textStyle(),
                color = MdStyle.Text,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

/** A chip's glyph at the painter's glyph size, tinted like the token text. */
@Composable
fun EntityChipGlyph(icon: ImageVector, tint: androidx.compose.ui.graphics.Color = MdStyle.ChipToken) {
    Box(Modifier.size(MdStyle.chipIconSize), contentAlignment = Alignment.Center) {
        Icon(icon, contentDescription = null, tint = tint, modifier = Modifier.size(MdStyle.chipIconSize))
    }
}

/**
 * EXP-920: the icon CONCEPT an entity-preview kind names
 * (`EntityPreview.ICON`, `packages/icons` `semantic`) as this build's glyph.
 * An explicit `when` — the registry generates concept ACCESSORS, not a
 * by-name lookup — locked by `EntityChipIconTest` so a concept the contract
 * adds can never fall through to null.
 */
fun entityConceptIcon(concept: String): ImageVector? = when (concept) {
    "ui-issue" -> ExpIcons.uiIssue
    "nav-boards" -> ExpIcons.navBoards
    "nav-actions" -> ExpIcons.navActions
    "nav-automations" -> ExpIcons.navAutomations
    "notification-issue-comment" -> ExpIcons.notificationIssueComment
    "coding-running" -> ExpIcons.codingRunning
    "settings-labels" -> ExpIcons.settingsLabels
    "settings-statuses" -> ExpIcons.settingsStatuses
    "nav-workflows" -> ExpIcons.navWorkflows
    "ui-device" -> ExpIcons.uiDevice
    "ui-avatar-placeholder" -> ExpIcons.uiAvatarPlaceholder
    "ui-repository" -> ExpIcons.uiRepository
    "ui-team" -> ExpIcons.uiTeam
    "ui-invite" -> ExpIcons.uiInvite
    "nav-notifications" -> ExpIcons.navNotifications
    "nav-support" -> ExpIcons.navSupport
    "ui-attach" -> ExpIcons.uiAttach
    "ui-checklist" -> ExpIcons.uiChecklist
    else -> null
}
